airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py

```diff
@@ -1154,17 +1158,24 @@ def add_to_outputs(
                     {{ '%}' }}
 
                     -- Delete records which are no longer active:
-                    -- The first subquery finds the most recent increment to the SCD table
-                    -- The second subquery finds, within that increment, the records which are still active
-                    -- We want to delete rows which are in that increment, but are not active
+                    -- This query is equivalent, but the left join version is more performant:
+                    -- delete from final_table where unique_key in (
+                    --     select unique_key from scd_table where 1 = 1 <incremental_clause(normalized_at, final_table)>
+                    -- ) and unique_key not in (
+                    --     select unique_key from scd_table where active_row = 1 <incremental_clause(normalized_at, final_table)>
+                    -- )
                     {{ delete_statement }} where {{ unique_key_reference }} in (
-                        select {{ unique_key }}
-                        from {{ '{{ this }}' }}
-                        where 1 = 1 {{ normalized_at_incremental_clause }}
-                    ) and {{ unique_key_reference }} not in (
-                        select {{ unique_key }}
+                        select distinct {{ unique_key }}
                         from {{ '{{ this }}' }}
-                        where {{ active_row_column_name }} = 1 {{ normalized_at_incremental_clause }}
+                        left join (
+                            select {{ unique_key }} as active_unique_key
+                            from {{ '{{ this }}' }}
+                            where {{ active_row_column_name }} = 1 {{ normalized_at_incremental_clause }}
+                        ) active_recent_scd_rows on {{ unique_key }} = active_unique_key
+                        where 1=1 {{ normalized_at_incremental_clause }}
+                        group by {{ unique_key }}
+                        having count(active_unique_key) = 0
```

Comment:
is it mandatory to use of `having` clause? I'm not sure about performance if it's not less efficient than other approaches using `where` if possible instead?

Reply:
running some manual performance tests; at first glance neither version has been significantly faster, but the non-having version does at least feel more readable :shrug: